### PR TITLE
Add an option to turn on reachability checks for assertions

### PR DESCRIPTION
**Pull request by zhassan-aws**
### Description of changes: 

A new option, `--assertion-reach-checks`, turns on reachability checks for assertions specified via the `assert!` macro. The results of reachability checks can aid in identifying vacuously passing properties that are reported as "SUCCESS" due to their unreachability (see #637). When this option is turned on, unreachable asserts would be reported as "UNREACHABLE".

### Call-outs:

1. The option currently turns on reachability checks for assertions specified via the `assert!` macro only. 
2. The `UNREACHABLE` result is only reported if the new output format (`--output-format regular`) is used. Without this option, the output would contain some raw unprocessed output that is not supposed to be user consumable.
3. Reachability checks are implemented via `__CPROVER_assert(false)` as opposed to the cleaner `__CPROVER_cover`. But using `__CPROVER_cover` is currently blocked by https://github.com/diffblue/cbmc/issues/6613. Because of using `__CPROVER_assert(false)`, the way to interpret the results of reachability checks is:
  - "SUCCESS": Assert is unreachable
  - "FAILURE": Assert is reachable

### Testing:

* How is this change tested? A standalone test and a cargo kani test were added 

* Is this a refactor change? No

### Checklist
- [X] Each commit message has a non-empty body, explaining why the change was made
- [X] Methods or procedures are documented
- [X] Regression or unit tests are included, or existing tests cover the modified code
- [X] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: src/kani-compiler/rustc_codegen_kani/src/overrides/hooks.rs [hunk 1]
```diff
@@ -169,18 +170,24 @@ impl<'tcx> GotocHook<'tcx> for Assert {
         let loc = tcx.codegen_span_option(span);
         let caller_loc = tcx.codegen_caller_span(&span);
 
+        let mut stmts: Vec<Stmt> = Vec::new();
+
+        if tcx.queries.get_check_assertion_reachability() {
+            // inject a reachability (cover) check to the current location
+            let reach_msg = format!("kani_reachability_check for {}", msg);
```
Comment (celinval) [inline src/kani-compiler/rustc_codegen_kani/src/overrides/hooks.rs:177]: Nit: Is this meant to be user visible?

**Review by celinval — CHANGES_REQUESTED** [review]
It looks good to me, but I would prefer blocking this last PR till we have the results parser changes in place. Till then, turning this on with basically fail any verification.

**Comment by zhassan-aws** [comment]
> I would prefer blocking this last PR till we have the results parser changes in place.

Sure, I'll add the parser changes to this PR.

File: src/kani-compiler/rustc_codegen_kani/src/overrides/hooks.rs (hunk 1 above)
Comment (zhassan-aws) [reply src/kani-compiler/rustc_codegen_kani/src/overrides/hooks.rs:177]: No, this is meant to make it easy for the CBMC result parser to match the reachability check with the assertion.

Comment (celinval) [reply src/kani-compiler/rustc_codegen_kani/src/overrides/hooks.rs:177]: Ok. In that case, I would replace the `for` by either `:` or `[]`:
```
let reach_msg = format!("[KANI_REACHABILITY_CHECK] {}", msg);
```
or
```
let reach_msg = format!("KANI_REACHABILITY_CHECK: {}", msg);
```

I think it makes it more clear that this is some sort of tag / category.

**Comment by zhassan-aws** [comment]
Added parser changes.

**Comment by zhassan-aws** [comment]
Addressed an issue that @celinval caught, which is that `kani` was terminating with a non-zero exit code even if the user-level checks were all successful. This happened because of the failure of the `assert(false)` statements that check reachability.

File: scripts/cbmc_json_parser.py [hunk 2]
```diff
@@ -469,6 +474,13 @@ def construct_property_message(properties):
 
     return output_message
 
+def count_failures(properties):
```
Comment (celinval) [inline scripts/cbmc_json_parser.py:477]: Don't we have this counter logic somewhere else where we print the execution summary? We should keep only one source of truth to the number of failures.

Comment (zhassan-aws) [reply scripts/cbmc_json_parser.py]: Deleted this function, and returned the number of failures from the functions that construct the summary message.

**Review by celinval — APPROVED** [review]
Can you please add a testcase or a line in the expected files that a reachable property is still reported as success? Maybe also add a testcase for the failure case too.

That can be in a follow up PR if you prefer.

Thanks

**Comment by zhassan-aws** [comment]
> Can you please add a testcase or a line in the expected files that a reachable property is still reported as success? Maybe also add a testcase for the failure case too.
>

Done.

**Pull request merged**